In [1]:
# imports
# best practice is to keep all the imports in one cell, at the top of your notebook or script

import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_squared_error, root_mean_squared_error


In [2]:
# Bring the prepared csv data in

df = pd.read_csv('data/cleanedSO.csv')

 Using some of the basic pandas tools, explore this dataset. Think back to descriptive statistics and meta data about the dataframe.

In [3]:
df.describe()

,YearsCode,Salary
count,21807.000000,21807.000000
mean,15.685055,90012.179599
std,9.926717,68987.446492
min,1.000000,0.024000
25%,8.000000,42560.000000
50%,13.000000,75000.000000
75%,20.000000,120000.000000
max,50.000000,498000.000000


In [4]:
df.head()

,Age,RemoteWork,EdLevel,YearsCode,Salary,primaryDB,primaryLang
0,25-34 years old,"Hybrid (some remote, some in-person)","Professional degree (JD, MD, Ph.D, Ed.D, etc.)",12,31360.0,other,C
1,35-44 years old,Remote,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",15,95200.0,other,JavaScript
2,35-44 years old,Remote,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",27,56000.0,BigQuery,Python
3,25-34 years old,Remote,Some college/university study without earning ...,7,110000.0,other,HTML/CSS
4,35-44 years old,"Hybrid (some remote, some in-person)","Professional degree (JD, MD, Ph.D, Ed.D, etc.)",32,166874.4,BigQuery,C#


In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 21807 entries, 0 to 21806
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Age          21807 non-null  str    
 1   RemoteWork   21807 non-null  str    
 2   EdLevel      21807 non-null  str    
 3   YearsCode    21807 non-null  int64  
 4   Salary       21807 non-null  float64
 5   primaryDB    21807 non-null  str    
 6   primaryLang  21807 non-null  str    
dtypes: float64(1), int64(1), str(5)
memory usage: 1.2 MB


#### Questions to consider:

1. Does anything jump out while looking at the descriptive statistics?<br>
A: The 'price' column has a very high mean and standard deviation compared to the median, which suggests that there are some outliers in the data. Additionally, the 'size' column has a wide range of values, which could also indicate the presence of outliers or a skewed distribution.

2. What columns can we use without manipulation?<br>
A: The 'size' column can be used without manipulation as it is already in a numerical format. However, the 'price' column may require some manipulation to handle the outliers and skewness in the data.

3. How are we going to handle the non-numerics?<br>
A: We can handle the non-numerics by either encoding them into numerical values or by dropping them if they are not relevant to our analysis. In this case, since we are only interested in predicting the price based on the size, we can drop any non-numeric columns that do not contribute to our model.

### Data Preparation

Remember those non-numeric variables? We need to deal with them. We will do that by creating [dummy variables](https://en.wikipedia.org/wiki/Dummy_variable_(statistics)) using the built-in pandas method [`pd.get_dummies()`](https://pandas.pydata.org/docs/reference/api/pandas.get_dummies.html)

In [9]:
# pd.get_dummies is smart enough to look at the datatypes and only use the fields 
# that are object fields.

df = pd.get_dummies(df)

In [10]:
# check the datatypes!
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 21807 entries, 0 to 21806
Data columns (total 40 columns):
 #   Column                                                                                      Non-Null Count  Dtype  
---  ------                                                                                      --------------  -----  
 0   YearsCode                                                                                   21807 non-null  int64  
 1   Salary                                                                                      21807 non-null  float64
 2   Age_18-24 years old                                                                         21807 non-null  bool   
 3   Age_25-34 years old                                                                         21807 non-null  bool   
 4   Age_35-44 years old                                                                         21807 non-null  bool   
 5   Age_45-54 years old                               

Note that after "dummying" we now have all numeric columns

### Variable Selection and TTS

(recall that TTS stands for train-test-split)

Statsmodels uses terms like endogenous and exogenous variables. scikit-learn uses different terminology, more in keeping with the predictive nature.

The dependent variable, or, what we are trying to predict, is always `y`.

The independent variables, or, what we are using to predict `y` is called `X`.

Note the `y` is lower case and the `X` is upper case. This is in following with matrix naming conventions. `X` is a matrix of independent variables and `y` is a vector for the dependent variable.

In [ ]:
# Set the target, or the y variable



In [ ]:
# Set the independent variable matrix, X



In [ ]:
# create your train-test splits



### Modeling

We will now instantiate, train, and use a model to make predictions.

scikit-learn as a large number of methods, which we will not exploring in their entirety here. We will focus on two methods:
1. [`.fit()`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression.fit)
2. [`.score()`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression.score)

`.fit()` is the method we will use to train our linear regression algorithm.

`.score()` will return the default scoring method, which, for linear regression is $r^2$.

In [ ]:
# instantiate
# this creates a linear regression object that we can train and score



We now need to train the model. This is the process of applying the data to linear regression linear equation. Recall the closed form solution for linear regression.

$\vec{b} = (X^TX)^{-1}X^Ty$

Where $\vec{b}$ are the coefficients we are training. This is done via the `.fit()` method.

In [ ]:
# Train the model with the correct data set



# Evaluation

We are going to use the scikit-learn metrics $r^2$ and mean squared error to evaluate how the model is performing.

In [ ]:
# use .score(). Recall this returns the default scoring method, which, for linear regression is r^2.


print(f'training r^2: {train_r2}')

Looking at that numbers what are your take-aways?





Now we'll look at mean squared error. In order to get this metric, we need to create our predictions using our trained model. This is done with the [`.predict()`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression.predict) method.

In [ ]:
# Get predictions



In [ ]:
# Use y_pred and y with the mean squared error method


print(f'training MSE: {train_mse}')

That number is ridiculously large. The reason is this method returns the *squared* error. This doesn't really translate well, so we'll take the square of it to get a more usable number. Another advantage is the units of the RMSE are the same as the target. In this instance, RMSE will tell us the average error in dollars, since that's what the target is.

In [ ]:
# use a better metric


train_rmse

Our $r^2$ is .2282. This is a very low number.

Our root mean squared error is 60,410. This can be read as the average salary prediction is off by $60k. This is not good, either.

#### Unseen data

Now that we have trained our model, lets see how it performs on unseen data.

In [ ]:
# Score with test data


print(f'training r^2: {test_r2}')

In [ ]:
# Get the RMSE for the test set

print(f'testing RMSE: {test_rmse}')

In [ ]:
# As  reminder
print(f'training r^2: {train_r2}')
print(f'training r^2: {test_r2}')

print(f'training RMSE: {train_rmse}')
print(f'testing RMSE: {test_rmse}')

So what can we conclude from this?

1. The model isn't good. At all.
2. There is an issue with the data (a minimum salary of .024?)

The good news is that it isn't overfit. We got an equally bad model with both our training and testing data